zenith_utils/src/postgres_backend.rs

```diff
@@ -0,0 +1,147 @@
+// Server-side Postgres connection, as limited as we need.
+
+use crate::pq_proto::{BeMessage, FeMessage, FeStartupMessage, StartupRequestCode};
+use anyhow::bail;
+use anyhow::Result;
+use bytes::Bytes;
+use log::*;
+use std::io;
+use std::io::{BufReader, BufWriter, Write};
+use std::net::TcpStream;
+
+pub trait Handler {
+    // handle single query
+    fn process_query(&mut self, pgb: &mut PostgresBackend, query_string: Bytes) -> Result<()>;
+    // called on startup packet receival, allows to process params
+    fn startup(&mut self, _pgb: &mut PostgresBackend, _sm: &FeStartupMessage) -> Result<()> {
+        Ok(())
+    }
+}
+
+pub struct PostgresBackend {
+    socket: TcpStream,
+    // replication.rs wants to handle reading on its own in separate thread, so
+    // wrap in Option to be able to take and transfer the BufReader. Ugly, but I
+    // have no better ideas.
+    stream_in: Option<BufReader<TcpStream>>,
+    stream: BufWriter<TcpStream>,
+    init_done: bool,
+}
+
+impl PostgresBackend {
+    pub fn new(socket: TcpStream) -> Self {
+        PostgresBackend {
+            stream_in: Some(BufReader::new(socket.try_clone().unwrap())),
+            stream: BufWriter::new(socket.try_clone().unwrap()),
+            socket: socket,
+            init_done: false,
```

Comment:
If the messages used don't overlap, maybe the startup sequence could just be a different codepath in `run()`

Reply:
You mean separating startup sequence into one more loop (the code section is mispointed, apparently)? I think it won't lead to less lines of code, there is little benefit there.